lib/vega_lite/data.ex

```diff
@@ -113,33 +122,43 @@ defmodule VegaLite.Data do
 
       Data.heatmap(data, x: "category", y: "score", color: "score", text: "category")
 
-  """
-  @spec heatmap(Table.Reader.t(), keyword()) :: VegaLite.t()
-  def heatmap(data, fields), do: heatmap(Vl.new(), data, fields)
-
-  @doc """
-  Same as heatmap/2, but takes a valid `VegaLite` specification as the first argument.
-
-  ## Examples
-
-      data = [
-        %{"category" => "A", "score" => 28},
-        %{"category" => "B", "score" => 55}
-      ]
+  With an existing VegaLite spec:
 
       Vl.new(title: "Heatmap", width: 500)
       |> Data.heatmap(data, x: "category", y: "score", color: "score", text: "category")
   """
   @spec heatmap(VegaLite.t(), Table.Reader.t(), keyword()) :: VegaLite.t()
-  def heatmap(vl, data, fields) do
+  def heatmap(vl \\ Vl.new(), data, fields) do
     for key <- [:x, :y], is_nil(fields[key]) do
       raise ArgumentError, "the #{key} field is required to plot a heatmap"
     end
 
-    opts = build_options(data, fields, &heatmap_defaults/2)
-    build_heatmap_layers(vl, data, opts)
+    heatmap_no_data(vl, data, fields, &heatmap_defaults/2)
+    |> attach_data(data, fields)
   end
 
+  defp heatmap_no_data(vl, data, fields, fun) do
+    cols = columns_for(data)
```

Comment:
Right now each `*_no_data` function calls `columns_for(data)` but we could also pass it as argument.